Rewrite pseudocolor render code to improve speed

pierstitus opened the pull request:
Most of the time was lost in checking the color cache, which for doubles has an almost infinite chance of cache hit misses anyway, so I replaced that by a look up table for the index in the color map.

It is substantially faster now, and the number of color map items doesn't really matter anymore. 

Also fixes http://hub.qgis.org/issues/12581

m-kuhn commented on python/core/raster/qgscolorrampshader.sip:
```diff
@@ -26,6 +26,14 @@ class QgsColorRampShader : QgsRasterShaderFunction
       bool operator<( const QgsColorRampShader::ColorRampItem& other ) const;
     };
 
+    /** Supported methods for color interpolation.
```
If you move the docs to the individual items they will have the context

```
/**
 * Main comment
 */
enum ColorRamp_TYPE
{
    INTERPOLATED, //!< Interpolates the color between...
...
```

m-kuhn commented on python/core/raster/qgscolorrampshader.sip:
```diff
@@ -42,9 +50,6 @@ class QgsColorRampShader : QgsRasterShaderFunction
     /** \brief Get the color ramp type as a string */
     QString colorRampTypeAsQString();
 
```
Don't remove methods in core classes. They can be deprecated and if possible routed to a corresponding replacement.
They will be removed with the next major release (QGIS 3)

pierstitus committed:
Rewrite pseudocolor render code to improve speed

Most of the time was lost in checking the color cache, so that is replaced
by a look up table for the index in the color map.

pierstitus commented:
Ok, implemented your comments. (and did an amend and force-push to keep history clean)

m-kuhn commented:
There seems to be something wrong looking at the test results:
http://dash.orfeo-toolbox.org/testDetails.php?test=40039155&build=225547

**Note** I did not look at the code concerning the algorithm.

pierstitus committed:
Fix sign of safety correction

pierstitus commented:
How beautiful to have testing! Small sign mistake, now fixed.

m-kuhn commented on src/core/raster/qgscolorrampshader.h:
```diff
@@ -130,7 +130,7 @@ class CORE_EXPORT QgsColorRampShader : public QgsRasterShaderFunction
     /** Maximum size of the color cache. The color cache could eat a ton of
      * memory if you have 32-bit data
      * @deprecated will be removed in QGIS 3.0 */
```
The doxygen comment still has the deprecation tag in it

pierstitus replied:
Yes, I removed the Q_DECL_DEPRECATED because it caused compiler warnings, because it's used in it get/set methods. other option would be to remove this variable and replace get/set methods by stubs, but I ought it to be of too minor importance to think too much about it, as this function/variable will probably never be used in its existence.

m-kuhn replied:
I don't think we deprecate private members. There is not much use for it since we have the full control over every usage anyway so instead of a two-step process (1. deprecate 2. remove) like for public members on which 3rd party code depends we can just remove it right away.

In such cases I normally just add a `// TODO QGIS 3: remove this` note.

FYI, you can also use

    Q_NOWARN_DEPRECATED_PUSH
    useDeprecatedMethodOrVariable
    Q_NOWARN_DEPRECATED_POP

which comes in handy when tests are written that checks if deprecated code still works until removed.

pierstitus replied:
Ok, changed deprecated to TODO.

pierstitus committed:
change comment to remove inadequate @deprecated

m-kuhn commented on src/core/raster/qgscolorrampshader.h:
```diff
@@ -119,28 +120,18 @@ class CORE_EXPORT QgsColorRampShader : public QgsRasterShaderFunction
-    QMap<double, QColor> mColorCache;
+    /** Look up table to speed up finding the right color.
+      * It is initialized on the first call to shade(). */
+    QList<int> mLUT;
```
a QVector should be even better in terms of performance.

pierstitus committed:
Change QList to QVector for speed increase

mLUT and mColorRampItemList are changed,
     for mLUT it shouldn't matter that much since an int is stored
     sequentially anyway, also in a QList,
     for mColorRampItemList it should make a difference,
     though it has to be converted from/to QList to keep API

pierstitus commented:
Change QList to QVector, according to the comment of @m-kuhn. I couldn't measure a significant speed increase with operf, but it feels a little bit faster.

It is anyway quite fast now, next chance for improvement in the render chain, could be QgsRasterProjector, as OTF projection really slows down a lot.

FYI. According to the [docs](http://doc.qt.io/qt-5/qvector.html#details) a QList stores also sequentially for small items, such as int, so for mLUT it shouldn't really matter.

m-kuhn commented:
Thanks a lot @pierstitus
